stan/math/opencl/matrix_cl.hpp

```diff
@@ -427,54 +424,47 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
   }
 
   /**
-   * Assign a \c matrix_cl to another
+   * Assign a matrix_cl from one place to another
    */
-  matrix_cl<T>& operator=(const matrix_cl<T>& a) {
-    check_size_match("assignment of (OpenCL) matrices", "source.rows()",
-                     a.rows(), "destination.rows()", rows());
-    check_size_match("assignment of (OpenCL) matrices", "source.cols()",
-                     a.cols(), "destination.cols()", cols());
-    if (a.size() == 0) {
-      return *this;
-    }
-    view_ = a.view();
+  template <typename Mat, require_matrix_cl_t<std::is_arithmetic, Mat>...>
+  matrix_cl<T>& operator=(Mat&& a) {
```

Comment:
Just updated the code, if you check out the code we just have one assign (the generic template one)

Reply:
We need to do very different thing in copy and move assignment operators. You are essentially making a move here. But copy assignment should copy the buffer. While this would be doable in single function with some ifs I think I prefere separate copy and move assignment operators. 

I'm ok with removing assignment of matrix of different type (which did the wrong thing anyway).